merkledag/ipld.md

```diff
@@ -0,0 +1,464 @@
+- IPLD paths MUST layer cleanly over UNIX and The Web (use `/`, have deterministic transforms for ASCII systems).
+- Given the wide success of JSON, a huge number of systems present JSON interfaces. IPLD MUST be able to import and export to JSON trivially.
+- The JSON data model is also very simple and easy to use. IPLD MUST be just as easy to use.
+- Definining new datastructures MUST be trivially easy. It should not be cumbersome -- or require much knowledge -- to experiment with new definitions on top of IPLD.
+- IPLD MUST be compatible with RDF and the proper W3C Semantic Web / Linked Data standards. We can achieve this easily through JSON-LD.
+- IPLD Serialized Formats (on disk and on the wire) MUST be fast and space efficient. (should not use JSON as the storage format, and instead use CBOR or similar formats)
+- IPLD cryptographic hashes MUST be upgradeable (use [multihash](https://github.com/jbenet/multihash))
+
+Some nice-to-haves:
+- IPLD SHOULD NOT carry over mistakes, e.g. the lack of integers in JSON.
+- IPLD SHOULD be upgradable, e.g. if a better on-disk format emerges, systems should be able to migrate to it and minimize costs of doing so.
+- IPLD objects SHOULD be able to resolve properties too as paths, not just merkle links.
+- IPLD Canonical Format SHOULD be easy to write a parser for.
+- IPLD Canonical Format SHOULD enable seeking without parsing full objects. (CBOR and Protobuf allow this).
+
+
+### Format Definition
+
+(**NOTE:** Here we will use both JSON and YML to show what formats look like. We explicitly use both to show equivalence of the object across two different formats.)
+
+At its core, IPLD Data Model "is just JSON" in that it (a) is also tree based documents with a few primitive types, (b) maps 1:1 to json, (c) users can use it through JSON itself. It "is not JSON" in that (a) it improves on some mistakes, (b) has an efficient serialized representation, and (c) does not actually specify a single on-wire format, as the world is known to improve.
+
+#### Basic Node
+
+Here is an example IPLD object in JSON:
+
+```json
+{
+  "name": "Vannevar Bush"
+}
+```
+
+Suppose it hashes to the multihash value `QmAAA...AAA`. Note that it has no links at all, just a string name value. But we are still be able to "resolve" the key `name` under it:
+
+```sh
+> ipld cat --json QmAAA...AAA
+{
+  "name": "Vannevar Bush"
+}
+
+> ipld cat --json QmAAA...AAA/name
+"Vannevar Bush"
+```
+
+And -- of course -- we are able to view it in other formats
+
+```sh
+> ipld cat --yml QmAAA...AAA
+---
+name: Vannevar Bush
+
+> ipld cat --xml QmAAA...AAA
+<!xml> <!-- todo -->
+<node>
+  <name>Vannevar Bush</name>
+</node>
+```
+
+#### Linking Between Nodes
+
+Merkle-Linking between nodes is the reason for IPLD to exist. A Link in IPLD is just an embedded node with a special format:
+
+```js
+{
+  "title": "As We May Think",
+  "author": {
+    "mlink": "QmAAA...AAA" // links to the node above.
```

Comment:
that will blow up verbosity and storage size. I was thinking, maybe we should drop `mlink` in favor of just `link` or `@link` (or the special char we go for). the link itself can denote what type of link it is:
- if it has no prefix (not a path, just a hash), it is a merkle-link.  (`"QmAAA...AAA"`)
- it may still have a prefix and be a merkle link (`"/ipfs/QmAAA...AAA/foo/bar"`)
- it may be another type of link ("/ipns/foo.com/a/b/c"`) but is not counted as immutable